module3/homework/advanced_calculator/advancedCalculator.cpp

```diff
@@ -0,0 +1,177 @@
+#include "advancedCalculator.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <iostream>
+#include <regex>
+
+auto addAction = [](double firstNumber, double secondNumber) {
+    return firstNumber + secondNumber;
+};
+
+auto substractAction = [](double firstNumber, double secondNumber) {
+    return firstNumber - secondNumber;
+};
+
+auto divideAction = [](double firstNumber, double secondNumber) {
+    return firstNumber / secondNumber;
+};
+
+auto multiplyAction = [](double firstNumber, double secondNumber) {
+    return firstNumber * secondNumber;
+};
+
+auto moduloAction = [](double firstNumber, double secondNumber) {
+    return (int)firstNumber % (int)secondNumber;
+};
+
+auto sqrtAction = [](double firstNumber, double secondNumber) {
+    return std::pow(firstNumber, 1 / secondNumber);
+};
+
+auto powerAction = [](double firstNumber, double secondNumber) {
+    return std::pow(firstNumber, secondNumber);
+};
+
+auto factorialAction = [](double firstNumber, double secondNumber) {
+    return std::tgamma(firstNumber + 1);
+};
+
+std::string eraseSpaces(std::string input)
+{
+    input.erase(std::remove_if(input.begin(),
+                               input.end(),
+                               [](unsigned char x) { return std::isspace(x); }),
+                input.end());
+
+    return input;
+}
+
+ErrorCode allowedCharacters(std::string input)
+{
+    const std::string ALLOWEDCHARACTERS = "+*/-%!^$1234567890.,";
+
+    auto foundBadCharacter = input.find_first_not_of(ALLOWEDCHARACTERS);
+    if (foundBadCharacter != std::string::npos) {
+        return ErrorCode::BadCharacter;
+    }
+
+    if (input.find(',') != std::string::npos) {
+        return ErrorCode::BadFormat;
+    }
+
+    return ErrorCode::OK;
+}
+
+ErrorCode allowedFormat(std::string input)
+{
+    std::regex patternUnary("(([-]?[0-9]+)|([-]?[0-9]+[/.][0-9]+))(!)");
+    std::regex patternBinary("(([-]?[0-9]+)|([-]?[0-9]+[/.][0-9]+))([-/+/*///^%/$])(([-]?[0-9]+)|([-]?[0-9]+[/.][0-9]+))");
+
+    std::smatch singleMatch;
+
+    if (!std::regex_match(input, singleMatch, patternUnary) && !std::regex_match(input, singleMatch, patternBinary)) {
+        return ErrorCode::BadFormat;
+    }
+    return ErrorCode::OK;
```

Comment:
regex is expensive. You can use other method to validate format

Reply:
I was learning two days how to use regex :D can i leave like that :)? btw. how can i do it in another way? Using if and else ?